### PR TITLE
Inline laser control

### PR DESCRIPTION
**Pull request by shitcreek**
### UPDATE
See lower comments

### Requirements

`LASER_FEATURE`

### Description

**DO NOT MERGE - MORE TESTING IS NEEDED**

This PR draft attempts to fix some of the issues with the current laser inline codes.
This PR will allow for the use of inline "mode" with normal always on `M3`

### Benefits

- Enabling `LASER_POWER_INLINE` changes the normal behavior of `M3` and `M5`.
- `M5` is effectively pointless as it no longer turns off the laser right away, instead it waits until the next move.
- `M5 I` does not turn the laser off right away either. Thus the laser stays on if there isn't another move that follows.
- - The way around it currently is to end every file with a `G1 S0...`, `M3 S0`, `G0...` or `G28` with `LASER_MOVE_G0_OFF` or `LASER_MOVE_G28_OFF` 
- G moves without `S` that's preceded by `M3` without an `I` eg `M3 S255` + `G1 X10` would produce nothing
- - This is because `M3` without an `I` does not set the inline value, only the laser value.

test-results01

Issues that this PR (for now at least) does not address:
- There is a dwell of the laser at the end of the file
- The 3rd last inline power setting takes the power setting of the 2nd last command (see **test-end**)

test-results02

Additionally
```
  static inline void set_enabled(const bool enable) {
    set_power(enable ? TERN(SPINDLE_LASER_PWM, (power ?: (unitPower ? upower_to_ocr(cpwr_to_upwr(SPEED_POWER_STARTUP)) : 0)), 255) : 0);
  }
```
The above implies that in case `unitPower ` exists, use `SPEED_POWER_STARTUP` or 0, but it should be using `unitPower` or `SPEED_POWER_STARTUP` instead.
```
static inline void set_enabled(const bool enable) {
    set_power(enable ? TERN(SPINDLE_LASER_PWM, (power ?: (unitPower ?: upower_to_ocr(cpwr_to_upwr(SPEED_POWER_STARTUP)))), 255) : 0);
  }
```

[laser-test-gcodes.zip](https://github.com/MarlinFirmware/Marlin/files/5564580/laser-test-gcodes.zip)

### Notes

I can get `M5` to turn off laser without a follow up move command, however, the brings back the issue of `G1` without `S` turning off the laser when using `M3` without `I`

### Configurations

[laser confg.zip](https://github.com/MarlinFirmware/Marlin/files/5564551/laser.confg.zip)

### Related Issues

#18965 and #20182

### CODE REVIEW

File: Marlin/src/feature/spindle_laser.h [hunk 1]
```diff
@@ -115,10 +111,16 @@ class SpindleLaser {
   FORCE_INLINE static void refresh() { apply_power(power); }
   FORCE_INLINE static void set_power(const uint8_t upwr) { power = upwr; refresh(); }
 
-  #if ENABLED(SPINDLE_LASER_PWM)
+  static inline void set_enabled(const bool enable) {
+    set_power(enable ? TERN(SPINDLE_LASER_PWM, (power ?: (unitPower ?: upower_to_ocr(cpwr_to_upwr(SPEED_POWER_STARTUP)))), 255) : 0);
```
Comment (zauguin) [inline Marlin/src/feature/spindle_laser.h]: `unitPower` is specified in the cutter unit and not always a ocr value, therefore I think this should be

    set_power(enable ? TERN(SPINDLE_LASER_PWM, (power ?: upower_to_ocr(unitPower ?: cpwr_to_upwr(SPEED_POWER_STARTUP))), 255) : 0);

instead.

**Comment by thinkyhead** [comment]
Merged the latest changes in and resolved a few conflicts in that process. Please test to see that nothing was lost in the merge and everything is functioning as expected.

File: Marlin/src/gcode/control/M3-M5.cpp [hunk 2]
```diff
@@ -85,56 +85,57 @@ void GcodeSuite::M3_M4(const bool is_M4) {
...
-
-  planner.synchronize();   // Wait for previous movement commands (G0/G0/G2/G3) to complete before changing power
+  planner.synchronize();   // Wait for previous movement commands (G0/G1/G2/G3) to complete before changing power
   cutter.set_direction(is_M4);
 
   #if ENABLED(SPINDLE_LASER_PWM)
     if (parser.seenval('O')) {
       cutter.unitPower = cutter.power_to_range(parser.value_byte(), 0);
       cutter.set_ocr_power(cutter.unitPower); // The OCR is a value from 0 to 255 (uint8_t)
+      TERN_(planner.laser_inline.status.isInline, inline_ocr_power(cutter.unitPower))
```
Comment (thinkyhead) [inline Marlin/src/gcode/control/M3-M5.cpp]: The first argument to `TERN_` has to be a configuration option, such as `LASER_POWER_INLINE`.

**Comment by shitcreek** [comment]
image
Latest codes yield the above. I've been able to address the aforementioned issues:
- There is a dwell of the laser at the end of the file
- The 3rd last inline power setting takes the power setting of the 2nd last command (see test-end)

However, the laser turns on briefly when hitting an endstop. I think I have a clue as to why, I'll keep at it.

**Comment by thinkyhead** [comment]
> Merge branch 'Inline-Laser' of https://github.com/shitcreek/Marlin in… 

This merge is bad and possibly produced code you would never really want. I will post the fixes that fit the code that resulted, but it's possibly going to be different from your intent, so please check my work!

**Comment by shitcreek** [comment]
Latest tweaks produce more desirable outcomes.
Now when using `LASER_POWER_INLINE`:
-`M3` (without `I`) will turn on briefly for time set by `LASER_PULSE` - uses `delay()` which might not be ideal
-`M3 I` will activate inline mode - laser will stay on until turned off by `M5`, `M3 S0`, `G0` or `G28` if using `LASER_MOVE_G0_OFF` / `LASER_MOVE_G28_OFF`
-`M5 I` isn't needed/used

Gcodes used:
Screen Shot 2021-04-02 at 8 46 04 PM
Expected results:
image
Actual results:
laser_gcode_test-actualResults
Notice: the 3rd to last `G1 Sxxx` uses the power setting of the 2nd to last `G1 Sxxx`

Also notice: laser pause/restarts between moves [testB.gcode]:
Screen Shot 2021-04-02 at 10 45 21 PM

[gcodes used](https://github.com/MarlinFirmware/Marlin/files/6253182/gcodes.zip)
[configs used](https://github.com/MarlinFirmware/Marlin/files/6253186/configs.zip)

Still need to test: 
- with `LASER_POWER_INLINE_CONTINUOUS`
- with `CUTTER_POWER_RELATIVE`
- without `LASER_POWER_INLINE`
- without `LASER_POWER_INLINE_TRAPEZOID`

**Pull request closed without merging**